contracts/mixins/ProtocolTokenHandler.sol

```diff
@@ -51,51 +65,75 @@ contract ProtocolTokenHandler is ProtocolTokenUser {
+     * @param _signs array includes signs from authorized signers
      */
-    function withdraw(bytes memory _data, bytes[] memory _signs) public {
-        _signs = new bytes[](requiredCount);
-        bytes32 _hash = keccak256(_data).toEthSignedMessageHash();
+    function withdraw(address _recipient, address _contractAddress, uint256 _amount, uint256 _nonce, bytes[] memory _signs) public {
+        require(_contractAddress == address(this), "contract address disagreement");
+        require(_amount > 0, "amount should larger than zero");
 
-        for (uint256 i = 0; i < requiredCount; i++) {
-            address signer = _hash.recover(_signs[i]);
-            require(isSigner[signer], "signer not authorized");
-            require(!verifiedSigners[signer], "signer verified");
-            verifiedSigners[signer] = true;
-        }
+        uint256 oldNonce = userNonce[msg.sender];
+        require(_nonce > oldNonce, "nonce smaller than last know nonce");
+        userNonce[msg.sender] = _nonce;
+
+        bytes32 _hash = keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", keccak256(abi.encodePacked(_recipient, _contractAddress, _amount, _nonce))));
+        
+        uint8 v;
+        bytes32 r;
+        bytes32 s;
+        address signer;
+        bytes memory sign;
+        bytes[] memory signs = new bytes[](requiredCount);
+        address[] memory verifiedSigners = new address[](requiredCount);
+
+        signs = _signs;
 
         for (uint256 i = 0; i < requiredCount; i++) {
-            verifiedSigners[_hash.recover(_signs[i])] = false;
-        }
+            sign = signs[i];
+            require(sign.length == 65, "length of sign is wrong");
 
-        (address recipient, address contractAddress, uint256 amount, uint256 nonce) = abi.decode(_data, (address, address, uint256, uint256));
+            assembly {
+                r := mload(add(sign, 0x20))
+                s := mload(add(sign, 0x40))
+                v := add(0x1b, byte(0, mload(add(sign, 0x60))))
+            }
 
-        require(!usedNonces[nonce], "nonce used");
-        usedNonces[nonce] = true;
+            signer = ecrecover(_hash, v, r, s);
+            require(isSigner[signer], "signer not authorized");
 
-        require(contractAddress == address(this), "contract address disagreement");
+            for (uint256 q = 0; q < i; q++) {
+                require(signer != verifiedSigners[q], "signer verified");
+            }
+            
+            verifiedSigners[i] = signer;
+        }
 
-        (address _protocolTokenAddress, bool success) = _withdrawProtocolToken(recipient, amount);
-        require(_protocolTokenAddress == protocolTokenAddress, "protocol token address disagreement");
-        if (success) {
-            protocolTokenPaid = protocolTokenPaid.add(amount);
+        uint256 balance = IERC20(protocolTokenAddress).balanceOf(address(this));
+        if (_amount > balance) {
+            _amount = balance;
```

Comment:
good thinking, but i think, it would actually be better to revert if the amount is bigger than the balance. the user will expect to get the full reward and not just part of it. it's better to wait for a contract refill instead of making partial payments.